.genai/io/capstone/rag/embeddings.py

```
@@ -0,0 +1,212 @@
+"""
+Embeddings wrappers for generating vector representations of code.
+
+Supports:
+- OpenAI (text-embedding-3-small)
+- Gemini (gemini-embedding-001)
+"""
+
+from typing import List, Protocol
+from abc import ABC, abstractmethod
+
+
+class EmbeddingsBase(ABC):
+    """Abstract base class for embeddings providers."""
+
+    @abstractmethod
+    def count_tokens(self, text: str) -> int:
+        """Count the number of tokens in a text string."""
+        pass
+
+    @abstractmethod
+    def truncate_to_tokens(self, text: str, max_tokens: int = None) -> str:
+        """Truncate text to fit within token limit."""
+        pass
+
+    @abstractmethod
+    def embed_text(self, text: str) -> List[float]:
+        """Generate embedding for a single text."""
+        pass
+
+    @abstractmethod
+    def embed_batch(self, texts: List[str]) -> List[List[float]]:
+        """Generate embeddings for multiple texts efficiently."""
+        pass
+
+    @property
+    @abstractmethod
+    def dimension(self) -> int:
+        """Return the embedding dimension."""
+        pass
+
+
+class GeminiEmbeddings(EmbeddingsBase):
+    """Wrapper for Google's Gemini embedding API (free tier available)."""
+
+    MODEL = "models/gemini-embedding-001"
+    MAX_TOKENS = 2048  # Conservative limit for Gemini
```

47 : [Warning] The `dimension` parameter in the `GeminiEmbeddings` constructor is not validated. The docstring states it supports `768, 1536, or 3072`, but the code accepts any integer. Passing an unsupported value will result in a runtime API error when `embed_text` or `embed_batch` is called, instead of failing early during object instantiation. It would be more robust to validate the `dimension` value within the `__init__` method.